File: data/harnesses/Base/go/secure/CWE-611_sonar_2.py/main.go
package main

import (
	"bytes"
	"encoding/xml"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
)

type Schema struct {
	Root xml.Name
}

func getRoot() (*Schema, error) {
	path := "resources/config.xsd"

	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, errors.New("OSError")
	}

	upper := bytes.ToUpper(raw)
	if bytes.Contains(upper, []byte("<!DOCTYPE")) || bytes.Contains(upper, []byte("<!ENTITY")) {
		return nil, errors.New("forbidden DTD or ENTITY found")
	}

	decoder := xml.NewDecoder(bytes.NewReader(raw))
	decoder.Strict = true
	decoder.AutoClose = xml.HTMLAutoClose
	decoder.Entity = xml.HTMLEntity

	var root xml.Name
	for {
		token, err := decoder.Token()
		if err != nil {
			if errors.Is(err, io.EOF) {
				break
			}
			return nil, errors.New("ValueError")
		}

		if se, ok := token.(xml.StartElement); ok {
			root = se.Name
			break
		}
	}

	if root.Local == "" {
		return nil, errors.New("no root element found")
	}

	schema := &Schema{
		Root: root,
	}

	return schema, nil
}

func main() {
	base := filepath.Join("Test", "CWE-611_sonar_2")
	resdir := filepath.Join(base, "resources")
	err := os.MkdirAll(resdir, 0755)
	if err != nil {
		fmt.Fprintf(os.Stderr, "MkdirAll failed: %v\n", err)
		os.Exit(1)
	}

	xsd := []byte(`<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema">
  <xs:element name="config" type="xs:string"/>
</xs:schema>`)
	err = os.WriteFile(filepath.Join(resdir, "config.xsd"), xsd, 0644)
	if err != nil {
		fmt.Fprintf(os.Stderr, "WriteFile failed: %v\n", err)
		os.Exit(1)
	}

	cwd, err := os.Getwd()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Getwd failed: %v\n", err)
		os.Exit(1)
	}

	err = os.Chdir(base)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Chdir failed: %v\n", err)
		os.Exit(1)
	}

	defer func() {
		chdirErr := os.Chdir(cwd)
		if chdirErr != nil {
			fmt.Fprintf(os.Stderr, "Chdir back failed: %v\n", chdirErr)
			os.Exit(1)
		}
	}()

	schema, err := getRoot()
	if err != nil {
		fmt.Fprintf(os.Stderr, "getRoot failed on valid XSD: %v\n", err)
		os.Exit(1)
	}
	if schema == nil {
		fmt.Fprintf(os.Stderr, "schema is nil\n")
		os.Exit(1)
	}
	if schema.Root.Local == "" {
		fmt.Fprintf(os.Stderr, "schema Root Local is empty\n")
		os.Exit(1)
	}

	dtd := []byte(`<!DOCTYPE a [<!ENTITY x "y">]><xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema"></xs:schema>`)
	err = os.WriteFile(filepath.Join("resources", "config.xsd"), dtd, 0644)
	if err != nil {
		fmt.Fprintf(os.Stderr, "WriteFile DTD failed: %v\n", err)
		os.Exit(1)
	}

	_, err = getRoot()
	if err == nil {
		fmt.Fprintf(os.Stderr, "expected error for DTD content, got nil\n")
		os.Exit(1)
	}

	malformed := []byte(`<xs:schema`)
	err = os.WriteFile(filepath.Join("resources", "config.xsd"), malformed, 0644)
	if err != nil {
		fmt.Fprintf(os.Stderr, "WriteFile malformed failed: %v\n", err)
		os.Exit(1)
	}

	_, err = getRoot()
	if err == nil {
		fmt.Fprintf(os.Stderr, "expected error for malformed XML, got nil\n")
		os.Exit(1)
	}

	fmt.Println("All checks passed")
}

File: data/harnesses/Base/cpp/secure/CWE-611_sonar_2.py/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <algorithm>
#include <stdexcept>
#include <functional>
#include <filesystem>

namespace etree {
    class XMLParser {
    public:
        XMLParser(bool resolve_entities, bool no_network, bool load_dtd, bool huge_tree) {}
    };

    class Element {};

    class XMLSchema {
    public:
        XMLSchema(Element* root) {}
    };

    class Document {
    public:
        Element* getroot() {
            return new Element();
        }
    };

    Document parse(const std::string& path, XMLParser& parser) {
        std::ifstream file(path, std::ios::binary);
        if (!file) {
            throw std::runtime_error("Failed to parse");
        }
        return Document();
    }
}

etree::XMLSchema get_root() {
    std::string path = "resources/config.xsd";
    std::vector<char> raw;

    try {
        std::ifstream file(path, std::ios::binary | std::ios::ate);
        if (!file) {
            throw std::runtime_error("Failed to open file");
        }
        std::streamsize size = file.tellg();
        file.seekg(0, std::ios::beg);
        raw.resize(size);
        if (!file.read(raw.data(), size)) {
            throw std::runtime_error("Failed to read file");
        }
    } catch (const std::exception& e) {
        throw std::runtime_error("OSError");
    }

    std::string upper(raw.begin(), raw.end());
    std::transform(upper.begin(), upper.end(), upper.begin(), ::toupper);

    if (upper.find("<!DOCTYPE") != std::string::npos || upper.find("<!ENTITY") != std::string::npos) {
        throw std::invalid_argument("ValueError");
    }

    try {
        etree::XMLParser parser(false, true, false, false);
        etree::Document treexsd = etree::parse(path, parser);
        etree::Element* rootxsd = treexsd.getroot();
        return etree::XMLSchema(rootxsd);
    } catch (const std::exception& e) {
        throw std::invalid_argument("ValueError");
    }
}

bool assert_raises(std::function<void()> func, const std::string& expected) {
    try {
        func();
        return false;
    } catch (const std::exception& e) {
        return std::string(e.what()).find(expected) != std::string::npos;
    }
}

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void write_file(const std::string& path, const std::string& content) {
    std::ofstream file(path, std::ios::binary);
    if (!file) {
        throw std::runtime_error("Failed to write file: " + path);
    }
    file.write(content.data(), content.size());
}

int main() {
    namespace fs = std::filesystem;

    std::string base = "Test/CWE-611_sonar_2";
    std::string resdir = base + "/resources";

    try {
        fs::create_directories(resdir);
    } catch (const std::exception& e) {
        fail("Failed to create directories: " + std::string(e.what()));
        return 1;
    }

    std::string xsd = "<xs:schema xmlns:xs=\"http://www.w3.org/2001/XMLSchema\">\n"
                      "  <xs:element name=\"config\" type=\"xs:string\"/>\n"
                      "</xs:schema>";

    write_file(resdir + "/config.xsd", xsd);

    fs::path cwd = fs::current_path();
    fs::current_path(base);

    int failures = 0;

    try {
        etree::XMLSchema schema = get_root();
        expect_true(true, "get_root() should succeed for valid XSD");
        
        std::string dtd = "<!DOCTYPE a [<!ENTITY x 'y'>]><xs:schema xmlns:xs='http://www.w3.org/2001/XMLSchema'></xs:schema>";
        write_file("resources/config.xsd", dtd);
        bool raises_dtd = assert_raises([]() { get_root(); }, "ValueError");
        expect_true(raises_dtd, "get_root() should raise ValueError for DTD content");

        write_file("resources/config.xsd", "<xs:schema");
        bool raises_malformed = assert_raises([]() { get_root(); }, "ValueError");
        expect_true(raises_malformed, "get_root() should raise ValueError for malformed XML");

    } catch (const std::exception& e) {
        fail("Unexpected exception: " + std::string(e.what()));
        failures++;
    }

    fs::current_path(cwd);

    try {
        fs::remove_all("Test");
    } catch (...) {}

    if (failures > 0) {
        return 1;
    }

    return 0;
}